Results Cannot Contain Optional Collections

This Slice definition produces code that doesn't compile:
```
struct TestingStuff {
    x: Result<bool, Sequence<int32>?>
}
```
It gives:
```
error CS8651: It is not legal to use nullable reference type 'IList<int>?' in an as expression; use the underlying type 'IList<int>' instead.
```

--- a/tests/ZeroC.Slice.Tests/ResultTests.slice
+++ b/tests/ZeroC.Slice.Tests/ResultTests.slice
@@ -21,3 +21,20 @@ compact enum StringOptInt32Result {
 compact struct StringOptInt32ResultHolder {
     value: Result<string, int32?>
 }
+
+// Just to test that the generated code compiles.
+struct OptionalCollectionResults {
+    r1: Result<string, Sequence<Dictionary<string, bool?>>>
+    r2: Result<string, Sequence<Dictionary<string, bool>?>>
+    r3: Result<string, Sequence<Dictionary<string, bool>>?>
+    r4: Result<string, Sequence<Dictionary<string, bool>>>?
+
+    r5: Result<string, Sequence<Dictionary<string, bool?>?>?>?
+
+    r6: Result<Sequence<Dictionary<string, bool?>>, string>
+    r7: Result<Sequence<Dictionary<string, bool>?>, string>
+    r8: Result<Sequence<Dictionary<string, bool>>?, string>
+    r9: Result<Sequence<Dictionary<string, bool>>, string>?
+
+    r10: Result<Sequence<Dictionary<string, bool?>?>?, string>?
+}

--- a/tools/slicec-cs/src/decoding.rs
+++ b/tools/slicec-cs/src/decoding.rs
@@ -390,7 +390,7 @@ fn decode_result_field(type_ref: &TypeRef, namespace: &str, encoding: Encoding) 
 
     // TODO: it's lame to have to do this here. We should provide a better API.
     if matches!(type_ref.concrete_type(), Types::Sequence(_) | Types::Dictionary(_)) {
-        write!(decode_func, " as {}", type_ref.field_type_string(namespace, false));
+        write!(decode_func, " as {}", type_ref.field_type_string(namespace, true));
     }
 
     if type_ref.is_optional {

